server/src/main/java/org/opensearch/cluster/routing/RoutingNode.java

```diff
@@ -48,63 +49,140 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 /**
  * A {@link RoutingNode} represents a cluster node associated with a single {@link DiscoveryNode} including all shards
  * that are hosted on that nodes. Each {@link RoutingNode} has a unique node id that can be used to identify the node.
  */
 public class RoutingNode implements Iterable<ShardRouting> {
 
+    static class BucketedShards implements Iterable<ShardRouting> {
+        private final Tuple<LinkedHashMap<ShardId, ShardRouting>, LinkedHashMap<ShardId, ShardRouting>> shardTuple; // LinkedHashMap to
+                                                                                                                    // preserve order
+
+        BucketedShards(LinkedHashMap<ShardId, ShardRouting> primaryShards, LinkedHashMap<ShardId, ShardRouting> replicaShards) {
+            this.shardTuple = new Tuple(primaryShards, replicaShards);
+        }
+
+        public boolean isEmpty() {
+            return this.shardTuple.v1().isEmpty() && this.shardTuple.v2().isEmpty();
+        }
+
+        public int size() {
+            return this.shardTuple.v1().size() + this.shardTuple.v2().size();
+        }
+
+        public boolean containsKey(ShardId shardId) {
+            return this.shardTuple.v1().containsKey(shardId) || this.shardTuple.v2().containsKey(shardId);
+        }
+
+        public ShardRouting get(ShardId shardId) {
+            if (this.shardTuple.v1().containsKey(shardId)) {
+                return this.shardTuple.v1().get(shardId);
+            }
+            return this.shardTuple.v2().get(shardId);
+        }
+
+        public ShardRouting add(ShardRouting shardRouting) {
```

Comment:
Nitpick, but I recommend naming this `put` since it is a convenience overload of the other put method.